newrelic-agent/src/main/java/com/newrelic/agent/config/ParsedConfigValue.java

```diff
@@ -0,0 +1,105 @@
+package com.newrelic.agent.config;
+
+import com.newrelic.agent.Agent;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.regex.Pattern;
+
+import static com.newrelic.agent.config.BaseConfig.LIST_ITEM_SEPARATOR;
+
+/**
+ * Any text config value has the potential to be interpreted as more than
+ * one data type.
+ * <p>
+ * This class allows for a text config value to be accessed
+ * as a type it can validly be coerced into, or accessed as a String.
+ * <p>
+ * This class also acts as a wrapper around simple data types that may be
+ * used in config - Boolean and Integer.
+ * <p>
+ * Parsing is carried out when config is loaded rather than at the time of access
+ * to avoid parsing more than once.
+ * <p>
+ * Retrieving a value with a null default value or without a default will return the original data type of
+ * the config - this will cause a {@link ClassCastException} if you attempt to retrieve to an incompatible
+ * type.
+ * <p>
+ * Retrieving a value with a default will attempt to return the config value coerced
+ * into the type of the default, if possible. If the defined value cannot be coerced
+ * into the type of the default value, then the default will be returned.
+ * <p>
+ * Be wary of coercing into Lists and Maps. Elements, keys and values will be Strings,
+ * but treating them as any other type will only fail when accessing individual
+ * members, not when retrieving the parsed value.
+ */
+public class ParsedConfigValue {
+
+    private static final Pattern MAP_TYPE_VALUE = Pattern.compile(".+:.+");
+    private final Object originalValue;
+    private List<String> valueAsList;
+    private Map<String, String> valueAsMap;
+    private Boolean valueAsBoolean;
+    private Integer valueAsInteger;
+
+    public ParsedConfigValue(Object value) {
```

Comment:
Really this is just pulling together existing functionality into one place. Basic idea is that the value is interpreted in as many ways as possible at time of construction, and can then be retrieved as any of the valid types by specifying a default of the required type